Document pattern-matching implementation (#5846)
carbon-language/carbon-lang: toolchain/docs/check/pattern_matching.md

# Pattern matching

<!--
Part of the Carbon Language project, under the Apache License v2.0 with LLVM
Exceptions. See /LICENSE for license information.
SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-->

<!-- toc -->

## Table of contents

-   [Glossary](#glossary)
-   [Overview](#overview)
-   [Pattern insts](#pattern-insts)
-   [Inst ordering](#inst-ordering)
-   [Parser-driven pattern block pushing](#parser-driven-pattern-block-pushing)
-   [Function parameters](#function-parameters)
    -   [`Call` parameters and arguments](#call-parameters-and-arguments)
    -   [Caller and callee matching](#caller-and-callee-matching)
    -   [The return slot](#the-return-slot)

<!-- tocstop -->

## Glossary

> A _full-pattern_ is a pattern that is not a subpattern of an enclosing pattern.

Reviewer: This seems different from the [definition in variadics](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/variadics.md):

> A _full pattern_ consists of an optional deduced parameter list, a pattern, and
an optional return type expression.

Could that same definition be used here, or is this using the same term for subtly different meanings?

As an aside, noting the main [pattern matching design](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/pattern_matching.md#overview) defines scrutinee, perhaps it should define for "full pattern" too?

Author: Good point! I've moved this definition and the one in `variadics.md` to `design/pattern_matching.md`.

The _scrutinee_ of a pattern is the value that a pattern is being matched

> against.

Reviewer: I'm used to glossaries being either separate or at the end of a document, rather than the start (particularly above the overview here, which I think our other documents put first).

But given you're only defining two terms, had you considered defining them at first use instead of at the top? i.e., giving a term a definition in the context it's used can often make it easier to remember the definition. I see you do this in at least some other cases, such as "pattern insts" just below, on line 45, or "pattern type" after that, on line 58. Why the different approach for these two terms?

Author: > Why the different approach for these two terms?

It's basically an accident of how I drafted this. I agree folding the definition into the body is better, but this is moot now that both terms are defined in the design docs.

## Overview

The SemIR for a pattern-matching operation is emitted in three steps:

1. Traverse the parse tree of the pattern to emit SemIR that abstractly
   describes the pattern.
2. Traverse the parse tree of the scrutinee expression to emit SemIR that
   evaluates it.

> 3. Traverse the pattern SemIR from step 1 (sometimes in conjunction with the

Reviewer: Have you considered naming these steps? You keep referring back to them in this document, but names might be easier... As it is, I find myself scrolling up on references to try to remember what step you're referring to when you say, for example, "steps 2 or 3".

For example, "1. Pattern: ...", "2. Scrutinee: ...", "3: Match: ...". Then you might say "The SemIR emitted in the 'pattern' step" instead of "The SemIR emitted in step 1"

Author: Oh, that's much better. Done!

   scrutinee SemIR) to emit SemIR that actually performs pattern matching.

> ## Pattern insts

Reviewer: For documentation, what would you think of writing out abbreviations, such as "instruction" instead of "inst"? In our [idioms](https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/docs/idioms.md#abbreviations-used-in-the-code-aka-carbon-abbreviation-decoder-ring), I see "abbreviations are typically only used in code, not comments".

Author: Interesting. In my head "inst" has become a term of art in its own right, like "impl", but it looks like our docs mostly don't agree with me. Fixed.

Reviewer: FWIW, I say `impl` mainly because that's the keyword, and I think it maps to a non-concise concept (I don't have a simple, unambiguous english term in my head).

insts that describe the pattern itself. For example, given the pattern
`(x: i32, y:i32)`, step 1 might emit the following SemIR:

```
%x.patt: %pattern_type.7ce = binding_pattern x [concrete]
%y.patt: %pattern_type.7ce = binding_pattern y [concrete]
%.loc4_21: %pattern_type.511 = tuple_pattern (%x.patt, %y.patt) [concrete]
```

Pattern insts do not represent executable code, and are generally ignored during
lowering. Instead, they descriptively represent the pattern itself as a kind of
constant value, and their primary consumer is step 3. The type of a pattern inst
is a _pattern type_, which is represented by a `pattern_type` inst. For example,
the `constants` block might define the types in the above SemIR like so:

```
%i32: type = class_type @Int, @Int(%int_32) [concrete]
%pattern_type.7ce: type = pattern_type %i32 [concrete]
%tuple.type: type = tuple_type (%i32, %i32) [concrete]
%pattern_type.511: type = pattern_type %tuple.type [concrete]
```

We can read this as saying that the type of `%x.patt` and `%y.patt` is "pattern
that matches an `i32` scrutinee", and the type of `%.loc4_21` is "pattern that
matches a `(i32, i32)` scrutinee".

> Insts that are not pattern insts are called _procedural insts_. All the insts

Reviewer: Are others using this "procedural inst" term for this purpose? This is the first I've heard it; glancing at [minutes](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.s437ufhxqbb), Discord, and code, I don't see this term being used. Where does the term come from?

Note, my own sense is that pattern instructions are special, but non-pattern instructions are things I'd bucket a few different ways, and I hesitate to call them broadly "procedural instructions" because I have a definition in my head of what a "procedure" is in programming that's more narrowly close to a function. If this is a new term you're proposing, had you considered just referring to it as "pattern instructions" and "non-pattern instructions"?

Author: Yeah, that's a fair point. "Non-pattern" is a little clunky in some places, but it works well enough.

emitted in steps 2 and 3 are procedural insts, but step 1 can emit procedural
insts as well. For example, in a pattern like `(x: i32, a + b)`, `i32` and
`a + b` are ordinary expressions, and so their SemIR must be emitted during the
initial traversal of the parse tree, as with any other expression.

All the pattern insts for a given full-pattern are grouped together in a
distinct block that contains no procedural insts. Consequently, `Check::Context`
maintains `pattern_block_stack` as a separate `InstBlockStack` for pattern
blocks, and provides separate methods like `AddPatternInst` for adding insts to
it.

## Inst ordering

The SemIR produced in the first two steps is (like most SemIR) generally in
bottom-up order, reflecting the order of the parse tree. However, the step 3
traversal is performed top-down, starting with the root inst of the pattern and

> traversing into its dependencies.

Reviewer: I'm a bit confused by the meaning of "bottom-up". For example, when we print the parse tree, we typically put the dependencies at the top (because we process postorder), and so a dependencies-first processing is top-down. Had you considered using tree traversal terms to reduce ambiguity?

```suggestion
The SemIR produced in the first two steps is (like most SemIR) generally in
post-order, reflecting the order of the parse tree. However, the step 3
traversal is performed pre-order, starting with the root inst of the pattern and
traversing into its dependencies.
```

In some cases it is necessary for step 1 to allocate insts that won't actually
be emitted until step 3, because they are responsible for performing pattern

> matching. When that happens, they are allocated but not added to a block, and
> their IDs are stored in the `Check::Context` (typically in a map from pattern
> inst IDs to the corresponding match inst IDs) so that they can be spliced into

Reviewer: Is this specifically talking about the below maps? Is there some other use-case, for what's atypical? If this is it, maybe:

```suggestion
matching. When that happens, they are allocated but not added to a block, and
their IDs are stored in the `Check::Context` (in one of the two below maps from pattern
inst IDs to the corresponding match inst IDs) so that they can be spliced into
```

Author: I don't think there are any atypical examples right now, but see below regarding future-proofing.

Reviewer: I'd really suggest "more may be added in the future" or similar instead of "typically", if you want more future-proofing. When you say "typically" and "for example", then I expect that there's _already_ an atypical case.

To try to explain how I think about this: the code may change, or it may not, it may change in some unpredictable way, or it can change alongside a doc update; trying to plan around what'll change in the future can be hard, unless you have a specific plan. However, if the documentation *already* says it's incomplete/unreliable, then it suggests people need to *already* read the code to understand what's going on. "Typically" in particular would make me hunt for what the "atypical" case is.

Author: That makes sense. How's this?

the current block at the appropriate point in step 3.

> For example:

Reviewer: Are these two examples from a longer list, or are these the two maps? i.e., could this say more precisely something like "The maps are:"

```suggestion
The maps are:
```

Author: At present I think those are the only two, but it seems possible that we'll add others in the future, and I'd like to keep these docs future-proof within reason.

signals from both the parser and the node stack to determine when to pop from
the pattern block stack.

In the case of `let` and `var` decls, this is fairly straightforward: the
beginning is marked by the `LetIntroducer` or `VarIntroducer` node, and the end
is marked by the `LetInitializer` or `VarInitializer`, or by the `VarDecl` in
the case of a `var` decl with no initializer. Similarly, the beginning of an
`impl forall` parameter list is marked by the `Forall` node, and the end is
marked by the `ImplDecl` or `ImplDefinitionStart`.

The case of a parameterized name (such as `Foo(x: i32)`) is more challenging.
The node immediately before the start of the full-pattern is an identifier, but
an identifier doesn't necessarily mark the start of a full-pattern. We've solved
that by having the parser mark identifier nodes that are followed by
full-patterns (using lookahead). Rather than use additional storage for what is
logically a single bit of data, we effectively smuggle that bit into the kind
enum by having separate node kinds `IdentifierNameBeforeParams` and
`IdentifierNameNotBeforeParams`.

If the parameterized name is a name qualifier (such as the first part of

> `Foo(X: i32).Bar(y: i32)`), the node immediately after it will be the qualifier

Reviewer: Typo?

```suggestion
`Foo(x: i32).Bar(y: i32)`), the node immediately after it will be the qualifier
```

Author: Yes, but I think the typo was omitting the `!`, not the capitalization.

mark the boundary: there is exactly one `ParamPattern` inst for each `Call`
parameter, which matches the entire corresponding `Call` argument. The
subpatterns of the `ParamPattern`s are matched on the callee side, and
everything above them is matched on the caller side. There are multiple kinds of
`ParamPattern` inst, which correspond to different ways of passing a parameter
(such as by reference or by value).

When performing callee-side pattern matching, we do not have an actual scrutinee
expression. Instead, for each `ParamPattern` inst we generate a corresponding
`Param` inst, which reads from the corresponding entry in the `Call` argument
list, and we use that as the scrutinee of the `ParamPattern`. Every
`ParamPattern` kind has a corresponding `Param` kind.

### The return slot

If a function has a declared return type, the function takes an additional
`Call` parameter, which points to the storage that should be initialized with
the return value. This `Call` parameter is represented as an `OutParamPattern`
inst with a `ReturnSlotPattern` inst as a subpattern. The `ReturnSlotPattern`
also represents the return type declaration itself, such as in

> `SemIR::FunctionFields`. The SemIR that matches these patterns consists of a

Reviewer: Why do you prefix some things with a namespace (`SemIR::` here, but also `Check::Context` since this is check documentation) but not others?

Author: I'm not sure why I qualified this one; I agree it's not necessary. I do think it's helpful to qualify `Check::Context` to help avoid potential confusion with `Parse::Context` -- I know it's not truly ambiguous because this is check documentation, but it could still be a temporary stumbling block for some readers.

`ReturnSlot` inst, which binds the special name `NameId::ReturnSlot` to the
`OutParam` inst representing the storage passed by the caller.

This structure is analogous to the handling of an ordinary by-value parameter,
which is represented in the `Call` parameters as a `ValueParamPattern` inst with
a `BindingPattern`, and in the pattern-matching SemIR as a `BindName` inst that
binds the parameter name to the `ValueParam` inst representing the argument
passed by the caller.

Note that if the return type does not have an in-place value representation
(meaning that the return value should not be passed in memory), these insts will

> all still be generated, but the `SemIR` for `return` statements will not access

Reviewer: Typo?

```suggestion
all still be generated, but the SemIR for `return` statements will not access
```

# Pattern matching

<!--
Part of the Carbon Language project, under the Apache License v2.0 with LLVM
Exceptions. See /LICENSE for license information.
SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-->

<!-- toc -->

## Table of contents

-   [Overview](#overview)
-   [Pattern instructions](#pattern-instructions)
-   [Instruction ordering](#instruction-ordering)
-   [Parser-driven pattern block pushing](#parser-driven-pattern-block-pushing)
-   [Function parameters](#function-parameters)
    -   [`Call` parameters and arguments](#call-parameters-and-arguments)
    -   [Caller and callee matching](#caller-and-callee-matching)
    -   [The return slot](#the-return-slot)

<!-- tocstop -->

## Overview

This document focuses on the implementation of pattern matching. See
[here](/docs/design/pattern_matching.md) for more on the design and fundamental
concepts.

The SemIR for a pattern-matching operation is emitted in three steps:

1. **Pattern:** Traverse the parse tree of the pattern to emit SemIR that
   abstractly describes the pattern.

> 2. **Scrutinee:**Traverse the parse tree of the scrutinee expression to emit

Reviewer: ```suggestion
2. **Scrutinee:** Traverse the parse tree of the scrutinee expression to emit
```

> make sense to put non-pattern instructions in the pattern block when they
> represent an expression that is part of the pattern. If so, substantial parts
> of this design might change. See
> [issue #5351](https://github.com/carbon-language/carbon-lang/issues/5351).

## Parser-driven pattern block pushing

At the same time as all of that, we have to manage the _pattern_ block stack as
well. We attempt to do this precisely rather than speculatively, by leveraging
the parser to precisely mark the nodes immediately before full-patterns, and
pushing the pattern block stack when we handle those nodes. We then rely on
signals from both the parser and the node stack to determine when to pop from
the pattern block stack.

In the case of `let` and `var` decls, this is fairly straightforward: the
beginning is marked by the `LetIntroducer` or `VarIntroducer` node, and the end
is marked by the `LetInitializer` or `VarInitializer`, or by the `VarDecl` in
the case of a `var` decl with no initializer. Similarly, the beginning of an
`impl forall` parameter list is marked by the `Forall` node, and the end is
marked by the `ImplDecl` or `ImplDefinitionStart`.

> The case of a parameterized name (such as `Foo(x: i32)`) is more challenging.

Reviewer: Should this match `Foo(X:! i32).Bar(y: i32)` below? i.e., one of:

```suggestion
The case of a parameterized name (such as `Foo(X:! i32)`) is more challenging.
```

```suggestion
The case of a parameterized name (such as `Bar(y: i32)`) is more challenging.
```
